x-pack/plugin/ml/src/test/java/org/elasticsearch/xpack/ml/dataframe/MappingsMergerTests.java

```diff
@@ -99,8 +100,145 @@ public void testMergeMappings_GivenIndicesWithDifferentMappingsButNoConflicts()
+        Map<String, Object> index1Mappings = Map.of("runtime", Map.of("field_1", "field_1_mappings"));
+        MappingMetadata index1MappingMetadata = new MappingMetadata("_doc", index1Mappings);
+
+        Map<String, Object> index2Mappings = Map.of("runtime", Map.of("field_1", "different_field_1_mappings"));
+        MappingMetadata index2MappingMetadata = new MappingMetadata("_doc", index2Mappings);
+
+        ImmutableOpenMap.Builder<String, MappingMetadata> mappings = ImmutableOpenMap.builder();
+        mappings.put("index_1", index1MappingMetadata);
+        mappings.put("index_2", index2MappingMetadata);
+
+        GetMappingsResponse getMappingsResponse = new GetMappingsResponse(mappings.build());
+
+        ElasticsearchStatusException e = expectThrows(ElasticsearchStatusException.class,
+            () -> MappingsMerger.mergeMappings(newSource(), getMappingsResponse));
+        assertThat(e.status(), equalTo(RestStatus.BAD_REQUEST));
+        assertThat(e.getMessage(),
+            containsString("cannot merge [runtime] mappings because of differences for field [field_1]; "));
+        assertThat(e.getMessage(), containsString("mapped as [different_field_1_mappings] in index [index_2]"));
+        assertThat(e.getMessage(), containsString("mapped as [field_1_mappings] in index [index_1]"));
+    }
+
+    public void testMergeMappings_GivenIndicesWithDifferentRuntimeFieldsButNoConflicts() {
+        Map<String, Object> index1Mappings = Map.of("runtime",
+            Map.of("field_1", "field_1_mappings", "field_2", "field_2_mappings"));
+        MappingMetadata index1MappingMetadata = new MappingMetadata("_doc", index1Mappings);
+
+        Map<String, Object> index2Mappings = Map.of("runtime",
+            Map.of("field_1", "field_1_mappings", "field_3", "field_3_mappings"));
+        MappingMetadata index2MappingMetadata = new MappingMetadata("_doc", index2Mappings);
+
+        ImmutableOpenMap.Builder<String, MappingMetadata> mappings = ImmutableOpenMap.builder();
+        mappings.put("index_1", index1MappingMetadata);
+        mappings.put("index_2", index2MappingMetadata);
+
+        GetMappingsResponse getMappingsResponse = new GetMappingsResponse(mappings.build());
+
+        MappingMetadata mergedMappings = MappingsMerger.mergeMappings(newSource(), getMappingsResponse);
+
+        Map<String, Object> mappingsAsMap = mergedMappings.getSourceAsMap();
+        assertThat(mappingsAsMap.size(), equalTo(2));
+        assertThat(mappingsAsMap.containsKey("dynamic"), is(true));
+        assertThat(mappingsAsMap.containsKey("runtime"), is(true));
+
+        assertThat(mappingsAsMap.get("dynamic"), is(false));
+
+        @SuppressWarnings("unchecked")
+        Map<String, Object> fieldMappings = (Map<String, Object>) mappingsAsMap.get("runtime");
+
+        assertThat(fieldMappings.size(), equalTo(3));
+        assertThat(fieldMappings.keySet(), containsInAnyOrder("field_1", "field_2", "field_3"));
```

Comment:
If you flipped the order of these assertions you'd get better diagnostics if the test ever failed.  As it stands you'll just get something like `expected:3 but was: 4`, which isn't enough to fix it.

There are several other places in the changes where the same thing applies.